### PR TITLE
Feature run E2E testing

### PR DESCRIPTION
**Pull request by metanerd**
#### Summary

- Updates mockgen
- Updates github lib
- Adds feature /e2e-test and /e2e-cancel as sketched [here](https://miro.com/app/board/o9J_lrJmbUI=/)

While deploying, please keep the revert button handy. 
Additional configurations can be found [here](https://git.internal.mattermost.com/mattermost/ci/mattermod/-/merge_requests/10) and need to be added by the release team

### CODE REVIEW

File: server/gitlab.go [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+				Value: info.WebappSHA,
+			},
+			{
+				Key:   envKeyBuildTag,
+				Value: info.BuildTag,
+			},
+			{
+				Key:   envKeyPRNumber,
+				Value: strconv.Itoa(info.TriggerPR),
+			},
```
Comment (saturninoabril) [inline server/gitlab.go:70]: I can't see where it includes optional pipeline variables such as for `MM_ENV`, `INCLUDE_FILE` and `EXCLUDE_FILE`. It's assigned to [info.EnvVars](https://github.com/mattermost/mattermost-mattermod/pull/282/files#diff-7047d41fe350835623854ec21ede87cc3c32f0f8842eba189be55bfb1d459bc6R180) but didn't see any loop/append to `Variables: []*gitlab.PipelineVariable`.

**Review by saturninoabril — CHANGES_REQUESTED** [review]
🎉 Yay, thank you Elisabeth! Looking forward to have this live! Overall looks good to me, except for some comments.

File: server/e2e_test_command.go [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+
+const (
+	e2eTestMsgCommenterPermission = "Looks like you don't have permissions to trigger this command.\n Only available for org members"
+	e2eTestMsgCIFailing           = "Command e2e-test requires PR checks to pass."
+	e2eTestMsgUnknownPRState      = "Failed to check if PR checks passed. I will not try to trigger e2e testing. Please retry in a bit."
+	e2eTestMsgPRInfo              = "Failed to get the PR information required to trigger testing. Please retry in a bit."
+	e2eTestMsgUnknownTriggerRepo  = "Looks like triggering e2e testing pipelines for this repo is not set up. \n Please contact your administrator of choice."
+	e2eTestMsgTrigger             = "Failed to trigger e2e testing pipeline."
+	e2eTestMsgCompanionBranch     = "Failed while trying to locate companion branch."
+	e2eTestMsgSameEnvs            = "Looks like a pipeline with the same environment variables is already running. \n Please cancel it first with /e2e-cancel or be more specific with environment variables."
```
[the remark this answers is not in the document]
Comment (metanerd) [reply server/e2e_test_command.go:23]: @mattermost/tech-writers could you please check these error messages?

File: server/gitlab.go [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+				Key:   "SHA_MATTERMOST_WEBAPP",
+				Value: info.WebappSHA,
+			},
+			{
+				Key:   "REF_MATTERMOST_SERVER",
+				Value: info.ServerBranch,
+			},
+			{
+				Key:   "SHA_MATTERMOST_SERVER",
+				Value: info.WebappSHA,
```
[the remark this answers is not in the document]
Comment (metanerd) [reply server/gitlab.go:61]: Thanks! Nice catch!

**Comment by cpanato** [comment]
one suggestion to help to review, can you split the PRs into smaller? like there are a lot of things that are not part of the main description in the pr.

I would suggest helping the review:
- go-Github update in a separate PR
- mockgen in a separate PR
- any other change that is not the main concern for the e2e feature in separated PR

and give a brief description of what this feature is to help the review.

**Comment by metanerd** [comment]
I tried to do this before. The process took too long with conflicting merges by other teams.

File: server/e2e_test_command.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (cwarnermm) [reply server/e2e_test_command.go:23]: @metanerd - We're new to these processes and don't currently have the knowledge/tool training to assist. Perhaps this an opportunity to pair up with QA and broaden our knowledge?

[the remark this answers is not in the document]
Comment (metanerd) [reply server/e2e_test_command.go:23]: Sure.

[the remark this answers is not in the document]
Comment (cwarnermm) [reply server/e2e_test_command.go:23]: @metanerd - I may have misinterpreted your request. Is the ask here to review the text of the error messages above to ensure they're clear, useful, and readable?

[the remark this answers is not in the document]
Comment (metanerd) [reply server/e2e_test_command.go:23]: Yes, exactly! If those error messages are understandable, not offensive, not intimidating, etc.

File: server/e2e_test_command.go [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/mattermost/mattermost-mattermod/model"
+	"github.com/mattermost/mattermost-server/v5/mlog"
+)
+
+const (
+	e2eTestMsgCommenterPermission = "Looks like you don't have permissions to trigger this command.\n Only available for org members"
```
Comment (cwarnermm) [inline server/e2e_test_command.go:16]:
```suggestion
	e2eTestMsgCommenterPermission = "You don't have permissions to trigger this command.\n It's only available for organization members."
```

File: server/e2e_test_command.go [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+	e2eTestMsgCIFailing           = "Command e2e-test requires PR checks to pass."
```
Comment (cwarnermm) [inline server/e2e_test_command.go:17]:
```suggestion
	e2eTestMsgCIFailing           = "The command /e2e-test requires all PR checks to pass."
```

File: server/e2e_test_command.go [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+	e2eTestMsgUnknownPRState      = "Failed to check if PR checks passed. I will not try to trigger e2e testing. Please retry in a bit."
```
Comment (cwarnermm) [inline server/e2e_test_command.go:18]:
```suggestion
	e2eTestMsgUnknownPRState      = "Failed to check whether PR checks passed. E2E testing isn't triggered. Please retry later."
```

File: server/e2e_test_command.go [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+	e2eTestMsgPRInfo              = "Failed to get the PR information required to trigger testing. Please retry in a bit."
```
Comment (cwarnermm) [inline server/e2e_test_command.go:19]:
```suggestion
	e2eTestMsgPRInfo              = "Failed to get the PR information required to trigger testing. Please retry later."
```

File: server/e2e_test_command.go [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+	e2eTestMsgUnknownTriggerRepo  = "Looks like triggering e2e testing pipelines for this repo is not set up. \n Please contact your administrator of choice."
```
Comment (cwarnermm) [inline server/e2e_test_command.go:20]:
```suggestion
	e2eTestMsgUnknownTriggerRepo  = "The ability to triggering E2E testing pipelines for this repository isn't set up. \n Please contact your system admin."
```

File: server/e2e_test_command.go [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+	e2eTestMsgTrigger             = "Failed to trigger e2e testing pipeline."
```
Comment (cwarnermm) [inline server/e2e_test_command.go:21]:
```suggestion
	e2eTestMsgTrigger             = "Failed to trigger E2E testing pipeline."
```

File: server/e2e_test_command.go [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+	e2eTestMsgCompanionBranch     = "Failed while trying to locate companion branch."
```
Comment (cwarnermm) [inline server/e2e_test_command.go:22]:
```suggestion
	e2eTestMsgCompanionBranch     = "Failed to locate companion branch."
```

File: server/e2e_test_command.go (hunk 2 above)
Comment (cwarnermm) [inline server/e2e_test_command.go:23]:
```suggestion
	e2eTestMsgSameEnvs            = "A pipeline with the same environment variables is already running. \n Please cancel it first with /e2e-cancel, or specify different environment variables."
```

**Review by cwarnermm — COMMENTED** [review]
Suggestions provided inline for user-facing messages. Please let me know if additional strings should be reviewed, or if you have any questions or concerns.

File: server/e2e_cancel.go [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+	case e2eCancelMsgCommenterPermission:
+		return commenterNoPermissions
+	case e2eCancelMsgFailedCancellation:
+		return "could not cancel"
+	case e2eCancelMsgNothingToCancel:
+		return "no pipeline to cancel"
+	default:
+		panic("unhandled error type")
+	}
+}
```
[the remark this answers is not in the document]
Comment (metanerd) [reply server/e2e_cancel.go:32]: I think this is for different types of users. One is a technical error message for developers and sysadmins and the other is a GitHub user comment. I would prefer to keep it that way.

File: server/e2e_test_command.go (hunk 8 above)
Comment (metanerd) [reply server/e2e_test_command.go:20]: Should it then be 
```suggestion
	e2eTestMsgUnknownTriggerRepo  = "The ability to trigger E2E testing pipelines for this repository isn't set up. \n Please contact a maintainer."
```

Comment (cwarnermm) [reply server/e2e_test_command.go:20]: Thanks for catching my typo! I agree with your updates.

File: server/e2e_test_command.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (metanerd) [reply server/e2e_test_command.go]: Thank you @cwarnermm !

[the remark this answers is not in the document]
Comment (saturninoabril) [reply server/e2e_test_command.go]: Thank you @cwarnermm!

**Review by saturninoabril — APPROVED** [review]
Thanks @metanerd! Great work!

**Review by furqanmlk — APPROVED** [review]
Great work @metanerd 👍 

Excited to see in action 🏃

**Comment by metanerd** [comment]
@cpanato Gentle reminder to review this PR.

File: cmd/mattermost-mattermod/main.go [hunk 12]
```diff
@@ -30,11 +30,11 @@ func main() {
 	config, err := server.GetConfig(configFile)
 	if err != nil {
 		mlog.Error("unable to load server config", mlog.Err(err), mlog.String("file", configFile))
-		os.Exit(1)
```
Comment (cpanato) [inline cmd/mattermost-mattermod/main.go:33]: i think this should stay as `os.Exit(1)`
same for the others

File: server/e2e_cancel_test.go [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+	"github.com/mattermost/mattermost-mattermod/server/mocks"
+	"github.com/stretchr/testify/require"
+)
+
+func TestHandleE2ECanceling(t *testing.T) {
+	ctrl := gomock.NewController(t)
+
+	const userHandle = "user"
+	const organization = "mattertest"
+	const prNumber = 123
```
Comment (cpanato) [inline server/e2e_cancel_test.go:25]:
```suggestion
	const (
	   userHandle = "user"
	   organization = "mattertest"
	   prNumber = 123
	)
```

**Review by cpanato — COMMENTED** [review]
looks good, a few nits.

need to open an MR here https://git.internal.mattermost.com/gitlab/argo-shared-cluster/-/blob/main/utilities/manifests/mattermod/configmap.yaml to add the new configuration.

after we merge this will run the release for mattermod to have the image available and then we can deploy that

one question: did you validate the code at least running in your local machine and connecting that to a test repo?

File: server/build_apps_test.go [hunk 14]
```diff
@@ -9,55 +9,55 @@ import (
 func TestGetExpectedArtifacts(t *testing.T) {
 	job := BuildAppJob{
 		JobName:           "a",
-		RepoName:          "honk",
+		RepoName:          "repo",
```
Comment (cpanato) [inline server/build_apps_test.go:12]: dont need to update those things

Comment (metanerd) [reply server/build_apps_test.go:12]: The test becomes hard to read for me. I already fell over it twice and was wondering what "honk" meant. Shouldn't the test be easy to read?

File: cmd/mattermost-mattermod/main.go (hunk 12 above)
Comment (metanerd) [reply cmd/mattermost-mattermod/main.go:33]: I followed this suggestion: https://github.com/mattermost/mattermost-mattermod/pull/275#discussion_r727288928

**Comment by metanerd** [comment]
> looks good, a few nits.
> 
> need to open an MR here https://git.internal.mattermost.com/gitlab/argo-shared-cluster/-/blob/main/utilities/manifests/mattermod/configmap.yaml to add the new configuration.
> 
> after we merge this will run the release for mattermod to have the image available and then we can deploy that
> 
> one question: did you validate the code at least running in your local machine and connecting that to a test repo?

I would need access to the argocd project, since I am not allowed to push to the project. 
Regarding your question, I ran a test server local with the tests. I did not connect the server to a test repo, because I would need a test organization and member setup.

File: server/build_apps_test.go (hunk 14 above)
Comment (cpanato) [reply server/build_apps_test.go:12]: ok

**Comment by cpanato** [comment]
> your question, I ran a test server local with the tests. I did not connect th

for that, you can use your own GitHub repository and setup that in your own repo under github.com/metanerd/some-repo

but ok, lets move forward

**Review by cpanato — APPROVED** [review]

**Comment by cpanato** [comment]
deployed

**Pull request merged**